src/statsdproto.rs

```diff
@@ -0,0 +1,385 @@
+    InvalidType,
+    #[error("invalid tag")]
+    InvalidTag,
+}
+
+/// A Tag is a key:value pair of metric tags
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+pub struct Tag {
+    name: Vec<u8>,
+    value: Vec<u8>,
+}
+
+/// Parsed gives an owned structure which represents a parsed statsd protocol
+/// unit which owns all of its fields. When parsing, no canonicalization is
+/// performed by default.
+#[derive(Debug, Clone)]
+pub struct Parsed {
+    name: Vec<u8>,
+    mtype: Type,
+    value: f64,
+    sample_rate: Option<f64>,
+    tags: Vec<Tag>,
+}
+
+fn parse_tags(input: &[u8]) -> Result<Vec<Tag>, ParseError> {
+    // Its impossible to have a tag of less than "a:b", also short circuit for
+    // no tags
+    match input.len() {
+        len if len == 0 => return Ok(vec![]),
+        len if len < 3 => return Err(ParseError::InvalidTag),
+        _ => (),
+    };
+
+    let mut tags: Vec<Tag> = Vec::new();
+    let mut scan = input;
+    loop {
+        let key_index_end = match memchr2(b':', b',', &scan[0..]) {
+            Some(i) if scan[i] == b':' => i,
+            Some(_) => return Err(ParseError::InvalidTag),
+            None => return Err(ParseError::InvalidTag),
+        };
+        let name = &scan[0..key_index_end];
+        scan = &scan[key_index_end + 1..];
+        match memchr2(b':', b',', scan) {
+            None => {
+                tags.push(Tag {
+                    name: name.to_vec(),
+                    value: scan.to_vec(),
+                });
+                return Ok(tags);
+            }
+            Some(value_index_end) if scan[value_index_end] == b',' => {
+                tags.push(Tag {
+                    name: name.to_vec(),
+                    value: scan[0..value_index_end].to_vec(),
+                });
+                scan = &scan[value_index_end + 1..];
+            }
+            Some(_) => return Err(ParseError::InvalidTag),
+        };
+    }
+}
+
+impl TryFrom<PDU> for Parsed {
+    type Error = ParseError;
+
+    fn try_from(pdu: PDU) -> Result<Self, Self::Error> {
+        let value = match lexical::parse::<f64, _>(pdu.value()) {
+            Ok(v) => v,
+            Err(_) => return Err(ParseError::InvalidValue),
+        };
+        let sample_rate = pdu
+            .sample_rate()
+            .map(|sr| match lexical::parse::<f64, _>(sr) {
+                Ok(v) if (v > 0.0 && v <= 1.0) => Ok(v),
+                Ok(_) => Err(ParseError::InvalidSampleRate),
+                Err(_) => Err(ParseError::InvalidSampleRate),
+            })
+            .transpose()?;
+        let mtype: Type = pdu.pdu_type().try_into()?;
+        let tags = pdu.tags().map(|v| parse_tags(v)).transpose()?;
+        Ok(Parsed {
+            name: pdu.name().to_vec(),
+            mtype: mtype,
+            value: value,
+            sample_rate: sample_rate,
+            tags: tags.unwrap_or_default(),
+        })
+    }
+}
+
+/// A PDU is an incoming protocol unit for statsd messages, commonly a
```

Comment:
Same docs nit for IDE users — short one-sentence summary up top split off by newlines is a bit easier to read in hovercards IMO (feel free to ignore if you prefer otherwise!)